boards/posix/nrf52_bsim/Kconfig.defconfig

```diff
@@ -38,5 +38,11 @@ config LOG_PROCESS_THREAD
 
 endif # LOG
 
+# Like other emulation environments that rely on host timekeeping (and
```

Comment:
Nono, the nrf52_bsim does not use the host time for anything. So the host scheduling should not affect in any way Zephyr's execution. What problem did you see? (I can help you debug it)

Reply:
Hm... how does it synthesize cycle counts out of the timer device then?  For example: if you spin waiting on the next tick, how does it know that time is passing?

Regardless, if you remove that patch you'll see failures out of sanitycheck when tickless is enabled.

Reply:
Oooh.... having typed that and walked away, I'm like 80% sure that's exactly the situation.  In "ticked" mode, the kernel responds with what amounts to an interrupt count when you ask it for the current time.  In tickless, it checks the current hardware counter for the delta from the last known tick instead.

If in native_posix, the "infinite clock rate" trick is going to mess that up.   In which case the workaround remains the same but I need to update the explanation to be correct :)  Let me dig a bit.

Reply:
> Presumably the clock looks frozen 

Exactly, the CPU/SW execution runs in 0 time (inside one simulation "delta cycle"), and simulated time only continues when `k_cpu_idle()` or `k_busy_wait()` are called.

Reply:
This fixes tests/kernel/tickless/tickless_concept
```diff
diff --git a/tests/kernel/tickless/tickless_concept/src/main.c b/tests/kernel/tickless/tickless_concept/src/main.c
index d0b39b9..636013e 100644
--- a/tests/kernel/tickless/tickless_concept/src/main.c
+++ b/tests/kernel/tickless/tickless_concept/src/main.c
@@ -33,7 +33,7 @@ static struct k_thread tdata[NUM_THREAD];
        do {                                   \
                u32_t t = k_uptime_get_32();   \
                while (t == k_uptime_get_32()) \
-                       posix_halt_cpu();      \
+                       k_busy_wait(50);      \
        } while (0)
 #else
 #define ALIGN_MS_BOUNDARY()                   \
```
Explanation:
The ALIGN_MS_BOUNDARY hack for the POSIX arch relied on having the ticker interrupt (between others) awaking the CPU.
But if you halt the CPU with the tick disabled, nothing would awake it for an arbitrary length of time. In that case, just replacing the CPU half with a small delay in each iteration of the loop is a good solution (50µs in this case, but any small, arbitrary number would do)

Reply:
Same fix for tests/kernel/sleep/
```diff
diff --git a/tests/kernel/sleep/src/main.c b/tests/kernel/sleep/src/main.c
index 8bebca0..19814af 100644
--- a/tests/kernel/sleep/src/main.c
+++ b/tests/kernel/sleep/src/main.c
@@ -73,7 +73,7 @@ static void align_to_tick_boundary(void)
        while (k_uptime_get_32() == tick) {
                /* Busy wait to align to tick boundary */
 #if defined(CONFIG_ARCH_POSIX)
-               posix_halt_cpu();
+               k_busy_wait(50);
 #endif
        }
```

Reply:
And more of the same for tests/kernel/sched/schedule_api
```diff
diff --git a/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c b/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c
index 6e51321..4668cf6 100644
--- a/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c
+++ b/tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c
@@ -104,7 +104,7 @@ void test_slice_reset(void)
                t32 = k_uptime_get_32();
                while (k_uptime_get_32() - t32 < HALF_SLICE_SIZE) {
 #if defined(CONFIG_ARCH_POSIX)
-                       posix_halt_cpu(); /*sleep until next irq*/
+                       k_busy_wait(50); /*sleep until next irq*/
 #else
                        ;
 #endif
```